ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,543 @@
+		if gr.incompleteCount == 0 {
+			gr.incompleteCount = dependencyFreeGroup
+		}
+	}
+
+	taskIdx := int64(-1)
+	defer atomic.StoreInt64(&taskIdx, tasksCount)
+	// create few go-routines to load asyncroniously the account data.
+	for i := 0; i < asyncAccountLoadingThreadCount; i++ {
+		go p.asyncPrefetchRoutine(&tasksQueue, &taskIdx, groupDoneCh)
+	}
+
+	// iterate on the transaction groups tasks. This array retains the original order.
+	completed := make(map[int]bool)
+	for i := 0; i < len(p.groups); {
+	wait:
+		incompleteCount := atomic.LoadInt64(&groupsReady[i].incompleteCount)
+		if incompleteCount > 0 || (incompleteCount != dependencyFreeGroup && !completed[i]) {
+			select {
+			case doneIdx := <-groupDoneCh:
+				if doneIdx < 0 {
+					doneIdx = -(doneIdx + 1)
+					// if there is an error, report the error to the output channel.
+					p.outChan <- loadedTransactionGroup{
+						err: groupsReady[doneIdx].err,
+					}
+					return
+				}
+				if doneIdx > i {
+					// mark future txn as ready.
+					completed[doneIdx] = true
+					goto wait
+				} else if doneIdx < i {
+					// it was already processed.
+					goto wait
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+		next := i
+		for ; next < len(p.groups); next++ {
+			if !completed[next] {
+				if next > i {
+					i = next
+					goto wait
+				}
+				// next == i
+			}
+
+			delete(completed, next)
+
+			// if we had no error, write the result to the output channel.
+			// this write will not block since we preallocated enough space on the channel.
+			p.outChan <- loadedTransactionGroup{
+				txnGroup:  p.groups[next],
+				accounts:  groupsReady[next].balances,
+				resources: groupsReady[next].resources,
+			}
+		}
+		// if we get to this point, it means that we have no more transaction to process.
+		break
+	}
+}
+
+func (gt *groupTask) markCompletionAcct(idx int, br loadedAccountDataEntry, groupDoneCh chan int) {
+	gt.balances[idx] = br
+	if atomic.AddInt64(&gt.incompleteCount, -1) == 0 {
+		// groupDoneCh reports back the group index that is complete and ready to be
+		// sent back on the channel.
+		groupDoneCh <- gt.groupTaskIndex
+	}
+}
+
+func (gt *groupTask) markCompletionResource(idx int, res loadedResourcesEntry, groupDoneCh chan int) {
+	gt.resources[idx] = res
+	if atomic.AddInt64(&gt.incompleteCount, -1) == 0 {
+		// groupDoneCh reports back the group index that is complete and ready to be
+		// sent back on the channel.
+		groupDoneCh <- gt.groupTaskIndex
+	}
+}
+
+func (gt *groupTask) markCompletionAcctError(err error, groupDoneCh chan int) {
+	for {
+		curVal := atomic.LoadInt64(&gt.incompleteCount)
+		if curVal <= 0 {
+			return
+		}
+		if atomic.CompareAndSwapInt64(&gt.incompleteCount, curVal, 0) {
+			gt.err = err
+			// groupDoneCh reports back the negative group index that could not
+			// complete. As mentioned above, we subtract additional 1 here
+			// in order to distinguish from the success case of index 0.
+			groupDoneCh <- -gt.groupTaskIndex - 1
```

Comment:
I guess another way without the negative value minus 1, etc, would be to set the highest bit on failure and then check for it on the other side? or have a struct that wraps two values, but I can tell you don't want a struct

Reply:
yes, setting the high bit is also an option. I've used the -1 on the fail case since it was the simpler implementation ( at least in my view.. ).